Civi/Core/Container.php

```diff
@@ -178,6 +179,16 @@ public function createContainer() {
       ))->setFactory('CRM_Utils_Cache::create');
     }
 
+    $container->setDefinition("cache.long", new Definition(
+      'CRM_Utils_Cache_Interface',
+      array(
+        array(
+          'name' => 'long',
+          'type' => array('SqlGroup', 'ArrayCache'),
```

Comment:
So 'default is

            'type' => array('*memory*', 'SqlGroup', 'ArrayCache'),

whereas 'long' is 

            'type' => array('*memory*', 'SqlGroup', 'ArrayCache'),

I feel like 'Redis' might be *memory* - even though it's potentially long-lived & not used at all for 'long'?

Where I wouldn't want to see Redis bypassed is Checks, Community Messages or JS_strings - that I know of - but they are all already 'default'

Reply:
I think Checks, Community Messages and JS_strings are currently 'long'

Reply:
hmm - they are def the sorts of things that would be good to be in Redis - ie they are retrieved every page load

Reply:
I'm trying to think through the logic - I mean what we want is to hit the quickest cache we can whenever we can without compromising data integrity.

I think we can assume that memory backed caches can easily store data for days or longer.

For sites with Redis I can't think of when you would want to be using an sql cache but I think think of a theoretical situation when you wouldn't want a memory 'layer' in front of the case - ie. if data is likely to be invalidated by another AJAX call

Reply:
1. Correction on `short`:  the short/default cache is really an alias for `CRM_Utils_Cache::singleton()` which simply obeys `CIVICRM_DB_CACHE_CLASS`. This is equivalent to `'type' => ['*memory*', 'ArrayCache']`.

2. I think it's a fair point that one would usually prefer Redis rather than SQL (when available). I'm not certain about memcache, though -- I had heard a vague rumor that memcached eviction behavior can be a bit harsher than Redis', but this isn't really something I know; the default tried to play-it-safe under the assumption that `long` may be storing large records. 

3. I'm happy to go with `Redis,SqlGroup,ArrayCache` (though I might have to tweak `CRM_Utils_Create::factory()` or `*memory*,SqlGroup,ArrayCache` if y'all really think it's better. It's greenfield, so the use-cases will adapt based on the default we choose.

Reply:
So I'm missing some terminolgy for 'prefer in-php caching if available, if not fall back to Redis' - which is making it hard for me to comment - what is the right term for that?

Reply:
I suspect the terms "prefer" and "fallback" are confusing because there are two different things.

1. *Cache factory/construction*: This the process of making an instance of `CacheInterface`. In this context, "prefer" or "fallback" refers to the idea of picking a specific object which implements a `CacheInterface` based on the current environment. (To wit: "*If there's Redis credentials, then instantiate `CRM_Utils_Cache_Redis`; otherwise, if there's SQL credentials, then instantiate `CRM_Utils_Cache_SqlGroup` class; otherwise, instantiate `CRM_Utils_Cache_ArrayCache`.*")
2. *Active cache tiers/hierarchy*: This is behavior of simultaneously storing the same record in multiple tiers. It can reduce read-latency, but it also has some complexity and penalties for cache-coherence. (To wit: *When writing, store this value in the SQL-cache and Redis and a local array. When reading, first check the local-array; then check Redis; then check SQL.*)

Much of the current code is about (1), and the preferences/fallbacks in this PR are about (1). But, in a separate conversation, we've talked about (2). Working on (2) is orthogonal to this PR. (A PR for (2) would look different -- perhaps adding a new class which uses the [decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern#Intent).)

Reply:
@totten ok so I guess for the concept of 'long' my reaction is still not to make it too easy for people to bypass Redis in favour of an sql cache when they don't have a clear reason to do so. I think by calling it 'long' people would say 'oh yeah this check should be stored for a long time'. I'm inclined for greenfield to favour memory caching & review if your vague rumour amounts to more than that

Reply:
(it's kinda hard to introduce a protocol for a theoretical long cache without working through an example usecase)

Reply:
"(To wit: When writing, store this value in the SQL-cache and Redis and a local array. When reading, first check the local-array; then check Redis; then check SQL.)"

In practice I would say only 2 not 3 cache layers but that doesn't change your point

Reply:
Fair point. Let's take a couple examples:

* (From yesterday) Coleman is working on an extension to display new-user tutorials in-app. The extension scans the entire `include_path` for JSON files with tutorial info; then it aggregates the data. On every page-view, it reads this JSON data to see if any tutorials match the current page-view. Scanning the filesystem on every page-view would be quite expensive, so instead it's cached. The `default`/`short` cache would work badly.
* (From a month or so back) In [afform](https://github.com/totten/afform/blob/master/CRM/Afform/AfformScanner.php), the list of available forms is built by scanning every extension and as well as `[civicrm.files]/afform`. There is an API for browsing/viewing/updating the form definitions. Scanning the filesystem on every Afform API call would also be expensive (though not quite as impactful on general end-users as the tutorial example). Instead, it's cached. The `default`/`short` cache would work badly.
* (I was chatting @bgm about this -- he said that something similar recently, but I don't remember the details.)

For the `tutorial` extension and @bgm's, they re-purposed an existing cache that uses `[*memory*,SqlGroup,ArrayCache]`. (Maybe `community_messages` or `js_strings`?) For `afform`, I added a new cache that also uses `[*memory*,SqlGroup,ArrayCache]`.

Talking through the examples, I think I agree with you now that it's better for `long` to use `[*memory*,SqlGroup,ArrayCache]`; that seems more typical. I'm pushing a commit to change it.